File: pong.go
// pong.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/exec"
	"time"
)

const (
	width  = 40
	height = 20
)

type Ball struct {
	x, y      float64
	dx, dy    float64
	speed     float64
	char      rune
}

type Game struct {
	balls       []Ball
	playerY     int
	aiY         int
	paddleH     int
	scorePlayer int
	scoreAI     int
	winScore    int
	gameOver    bool
	running     bool
	paddleSpeed int
}

func NewGame() *Game {
	g := &Game{
		paddleH:     4,
		winScore:    5,
		running:     true,
		paddleSpeed: 1,
	}
	g.playerY = height/2 - g.paddleH/2
	g.aiY = height/2 - g.paddleH/2
	g.initBalls()
	return g
}

func (g *Game) initBalls() {
	g.balls = []Ball{
		{float64(width / 2), float64(height / 2), -1, -0.5, 0.8, 'O'},
		{float64(width / 2), float64(height / 2), 1, 0.5, 0.9, 'o'},
	}
}

func clearScreen() {
	cmd := exec.Command("clear")
	cmd.Stdout = os.Stdout
	cmd.Run()
}

func (g *Game) drawBorder() {
	for x := 0; x < width; x++ {
		fmt.Printf("\033[%d;%dH#", 0, x)
		fmt.Printf("\033[%d;%dH#", height-1, x)
	}
	for y := 0; y < height; y++ {
		fmt.Printf("\033[%d;%dH#", y, 0)
		fmt.Printf("\033[%d;%dH#", y, width-1)
	}
	// centre line
	for y := 1; y < height-1; y++ {
		fmt.Printf("\033[%d;%dH|", y, width/2)
	}
}

func (g *Game) drawPaddles() {
	for i := 0; i < g.paddleH; i++ {
		fmt.Printf("\033[%d;%dH]", g.playerY+i, 2)
		fmt.Printf("\033[%d;%dH[", g.aiY+i, width-3)
	}
}

func (g *Game) drawBalls() {
	for _, b := range g.balls {
		if b.x > 0 && b.x < float64(width-1) && b.y > 0 && b.y < float64(height-1) {
			fmt.Printf("\033[%d;%dH%c", int(b.y), int(b.x), b.char)
		}
	}
}

func (g *Game) drawScore() {
	fmt.Printf("\033[%d;%dH%d : %d", 1, width/2-2, g.scorePlayer, g.scoreAI)
	if g.gameOver {
		winner := "Player"
		if g.scoreAI >= g.winScore {
			winner = "AI"
		}
		fmt.Printf("\033[%d;%dH%s WINS!", height/2, width/2-len(winner)/2, winner)
		fmt.Printf("\033[%d;%dHPress R to restart", height/2+1, width/2-8)
	}
}

func (g *Game) updateBalls() {
	if g.gameOver {
		return
	}
	for i := range g.balls {
		b := &g.balls[i]
		b.x += b.dx * b.speed
		b.y += b.dy * b.speed
		// top/bottom
		if b.y <= 1 || b.y >= float64(height-2) {
			b.dy *= -1
		}
		// left paddle
		if b.x <= 3 && int(b.y) >= g.playerY && int(b.y) < g.playerY+g.paddleH {
			b.dx *= -1
			b.x = 4
			b.speed *= 1.05
			if b.speed > 2.0 {
				b.speed = 2.0
			}
		}
		// right paddle
		if b.x >= float64(width-4) && int(b.y) >= g.aiY && int(b.y) < g.aiY+g.paddleH {
			b.dx *= -1
			b.x = float64(width - 5)
			b.speed *= 1.05
			if b.speed > 2.0 {
				b.speed = 2.0
			}
		}
		// scoring
		if b.x <= 0 {
			g.scoreAI++
			g.resetBall(i, 1)
			if g.scoreAI >= g.winScore {
				g.gameOver = true
			}
		} else if b.x >= float64(width-1) {
			g.scorePlayer++
			g.resetBall(i, -1)
			if g.scorePlayer >= g.winScore {
				g.gameOver = true
			}
		}
	}
}

func (g *Game) resetBall(idx int, dir int) {
	b := &g.balls[idx]
	b.x = float64(width / 2)
	b.y = float64(height / 2)
	b.dx = float64(dir) * (0.5 + rand.Float64()*0.5)
	b.dy = (rand.Float64() - 0.5) * 0.8
	b.speed = 0.8 + rand.Float64()*0.4
	if b.dy == 0 {
		b.dy = 0.3
		if rand.Float64() > 0.5 {
			b.dy = -0.3
		}
	}
}

func (g *Game) updateAI() {
	targetY := height / 2
	for _, b := range g.balls {
		if b.dx > 0 && b.x > float64(width/2) {
			targetY = int(b.y)
			break
		}
	}
	if targetY > g.aiY+g.paddleH/2+1 {
		g.aiY += g.paddleSpeed
	} else if targetY < g.aiY+g.paddleH/2-1 {
		g.aiY -= g.paddleSpeed
	}
	if g.aiY < 1 {
		g.aiY = 1
	}
	if g.aiY > height-g.paddleH-1 {
		g.aiY = height - g.paddleH - 1
	}
}

func (g *Game) movePlayer(dy int) {
	if g.gameOver {
		return
	}
	newY := g.playerY + dy
	if newY >= 1 && newY <= height-g.paddleH-1 {
		g.playerY = newY
	}
}

func (g *Game) restart() {
	g.playerY = height/2 - g.paddleH/2
	g.aiY = height/2 - g.paddleH/2
	g.scorePlayer = 0
	g.scoreAI = 0
	g.gameOver = false
	g.initBalls()
}

func (g *Game) render() {
	clearScreen()
	g.drawBorder()
	g.drawPaddles()
	g.drawBalls()
	g.drawScore()
}

func (g *Game) run() {
	go func() {
		for g.running {
			var b [1]byte
			os.Stdin.Read(b[:])
			switch b[0] {
			case 'q', 'Q':
				g.running = false
			case 'r', 'R':
				if g.gameOver {
					g.restart()
				}
			case 'w', 'W':
				g.movePlayer(-1)
			case 's', 'S':
				g.movePlayer(1)
			}
		}
	}()
	for g.running {
		g.updateAI()
		g.updateBalls()
		g.render()
		time.Sleep(16 * time.Millisecond)
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	fmt.Print("\033[?25l")
	game := NewGame()
	game.run()
}

File: pong.py
# pong.py
import curses
import random
import time

class Ball:
    def __init__(self, x, y, dx, dy, char='O'):
        self.x = x
        self.y = y
        self.dx = dx
        self.dy = dy
        self.char = char
        self.speed = 1.0

    def move(self):
        self.x += self.dx * self.speed
        self.y += self.dy * self.speed

    def bounce_x(self):
        self.dx *= -1

    def bounce_y(self):
        self.dy *= -1

    def speed_up(self, factor=1.05):
        self.speed = min(2.0, self.speed * factor)

class Pong:
    def __init__(self, stdscr):
        self.stdscr = stdscr
        curses.curs_set(0)
        self.stdscr.nodelay(1)
        self.stdscr.timeout(16)
        self.height, self.width = self.stdscr.getmaxyx()
        self.height = min(self.height, 24)
        self.width = min(self.width, 80)
        self.paddle_h = 4
        self.player_y = self.height // 2 - self.paddle_h // 2
        self.ai_y = self.height // 2 - self.paddle_h // 2
        self.score_player = 0
        self.score_ai = 0
        self.win_score = 5
        self.game_over = False
        self.paused = False
        self.running = True
        self.balls = []
        self.init_balls()
        self.paddle_speed = 1

    def init_balls(self):
        cx = self.width // 2
        cy = self.height // 2
        angles = [(-1, -0.5), (1, 0.5)]
        for i, (dx, dy) in enumerate(angles):
            b = Ball(cx, cy, dx, dy, 'O' if i == 0 else 'o')
            b.speed = 0.8 + random.random() * 0.4
            self.balls.append(b)

    def draw_border(self):
        for y in range(self.height):
            self.stdscr.addch(y, 0, '#')
            self.stdscr.addch(y, self.width-1, '#')
        for x in range(self.width):
            self.stdscr.addch(0, x, '#')
            self.stdscr.addch(self.height-1, x, '#')
        # center line
        for y in range(1, self.height-1):
            self.stdscr.addch(y, self.width//2, '|')

    def draw_paddles(self):
        for i in range(self.paddle_h):
            self.stdscr.addch(self.player_y + i, 2, ']')
            self.stdscr.addch(self.ai_y + i, self.width-3, '[')

    def draw_balls(self):
        for b in self.balls:
            if 0 < b.x < self.width-1 and 0 < b.y < self.height-1:
                self.stdscr.addch(int(b.y), int(b.x), b.char)

    def draw_score(self):
        self.stdscr.addstr(1, self.width//2 - 5, f"{self.score_player} : {self.score_ai}")
        if self.game_over:
            winner = "Player" if self.score_player >= self.win_score else "AI"
            self.stdscr.addstr(self.height//2, self.width//2 - len(winner)//2, f"{winner} WINS!")
            self.stdscr.addstr(self.height//2 + 1, self.width//2 - 8, "Press R to restart")

    def update_balls(self):
        if self.game_over or self.paused:
            return
        for b in self.balls:
            b.move()
            # top/bottom walls
            if b.y <= 1 or b.y >= self.height-2:
                b.bounce_y()
            # left paddle
            if b.x <= 3 and self.player_y <= b.y < self.player_y + self.paddle_h:
                b.bounce_x()
                b.x = 4
                b.speed_up()
            # right paddle
            if b.x >= self.width-4 and self.ai_y <= b.y < self.ai_y + self.paddle_h:
                b.bounce_x()
                b.x = self.width-5
                b.speed_up()
            # scoring
            if b.x <= 0:
                self.score_ai += 1
                self.reset_ball(b, 1)
                if self.score_ai >= self.win_score:
                    self.game_over = True
            elif b.x >= self.width-1:
                self.score_player += 1
                self.reset_ball(b, -1)
                if self.score_player >= self.win_score:
                    self.game_over = True

    def reset_ball(self, ball, direction):
        ball.x = self.width // 2
        ball.y = self.height // 2
        ball.dx = direction * (0.5 + random.random() * 0.5)
        ball.dy = (random.random() - 0.5) * 0.8
        ball.speed = 0.8 + random.random() * 0.4
        if ball.dy == 0:
            ball.dy = 0.3 if random.random() > 0.5 else -0.3

    def update_ai(self):
        # Simple AI: follow the ball that's closest to its side
        target_y = self.height // 2
        for b in self.balls:
            if b.dx > 0 and b.x > self.width // 2:
                target_y = b.y
                break
        if target_y > self.ai_y + self.paddle_h // 2 + 1:
            self.ai_y += self.paddle_speed
        elif target_y < self.ai_y + self.paddle_h // 2 - 1:
            self.ai_y -= self.paddle_speed
        self.ai_y = max(1, min(self.height - self.paddle_h - 1, self.ai_y))

    def move_player(self, dy):
        if self.game_over:
            return
        new_y = self.player_y + dy
        if 1 <= new_y <= self.height - self.paddle_h - 1:
            self.player_y = new_y

    def restart(self):
        self.player_y = self.height // 2 - self.paddle_h // 2
        self.ai_y = self.height // 2 - self.paddle_h // 2
        self.score_player = 0
        self.score_ai = 0
        self.game_over = False
        self.balls = []
        self.init_balls()

    def render(self):
        self.stdscr.clear()
        self.draw_border()
        self.draw_paddles()
        self.draw_balls()
        self.draw_score()
        self.stdscr.refresh()

    def run(self):
        while self.running:
            key = self.stdscr.getch()
            if key == ord('q') or key == ord('Q'):
                break
            if key == ord('r') or key == ord('R'):
                self.restart()
            if key == ord('w') or key == ord('W') or key == curses.KEY_UP:
                self.move_player(-1)
            elif key == ord('s') or key == ord('S') or key == curses.KEY_DOWN:
                self.move_player(1)
            self.update_ai()
            self.update_balls()
            self.render()
            time.sleep(0.016)

def main(stdscr):
    game = Pong(stdscr)
    game.run()

if __name__ == "__main__":
    curses.wrapper(main)
